### PR TITLE
Improved python3 compatibility of lib tests #1932

### PR DESCRIPTION
**Pull request by Onager**
Improved python3 compatibility of lib tests #1932

### CODE REVIEW

File: plaso/lib/objectfilter.py [hunk 1]
```diff
@@ -112,8 +113,15 @@ def __init__(self, brand, code):
 
 def GetUnicodeString(string):
   """Converts the string to Unicode if necessary."""
```
Comment (joachimmetz) [inline plaso/lib/objectfilter.py]: add args section, also maybe rename string argument to value?

File: plaso/lib/utils.py [hunk 2]
```diff
@@ -21,7 +21,7 @@ def IsText(bytes_in, encoding=None):
   sequence is not text, but a byte sequence.
 
   Args:
-    bytes_in (bytes): byte stream to examine.
+    bytes_in (bytes|str): byte stream to examine.
```
Comment (joachimmetz) [inline plaso/lib/utils.py]: why add str?

File: plaso/lib/utils.py [hunk 3]
```diff
@@ -32,19 +32,21 @@ def IsText(bytes_in, encoding=None):
   # Start with the assumption we are dealing with a text.
   is_ascii = True
 
+  if isinstance(bytes_in, py2to3.UNICODE_TYPE):
+    return is_ascii
```
Comment (joachimmetz) [inline plaso/lib/utils.py]: lets rename `is_ascii` to `is_text` seeing an ASCII Unicode string does not make sense to me

File: tests/lib/binary.py [hunk 4]
```diff
@@ -172,26 +172,26 @@ def testArrayOfUTF16StreamCopyToStringTable(self):
       self.assertEqual(string_table, expected_string_table)
 
   def testStringParsing(self):
```
Comment (joachimmetz) [inline tests/lib/binary.py]: name of function and docstring don't seem to match what is being tested

File: plaso/lib/utils.py (hunk 3 above)
Comment (joachimmetz) [reply plaso/lib/utils.py]: or define `is_ascii` after this check and return True

File: plaso/lib/binary.py [hunk 5]
```diff
@@ -28,44 +29,44 @@ def ByteStreamCopyToString(byte_stream, codepage='utf-8'):
...
   try:
-    string = byte_stream.decode(codepage)
+    string = codecs.decode(byte_stream, codepage)
   except UnicodeDecodeError:
     logging.warning(
         'Unable to decode {0:s} formatted byte stream.'.format(codepage))
-    string = byte_stream.decode(codepage, errors='ignore')
+    string = codecs.decode(byte_stream, codepage, errors='ignore')
 
   string, _, _ = string.partition('\x00')
```
Comment (joachimmetz) [inline plaso/lib/binary.py]: Wasn't split faster? Maybe change to `string = string.split('\x00')[0]`

File: plaso/lib/objectfilter.py [hunk 6]
```diff
@@ -110,11 +111,25 @@ def __init__(self, brand, code):
...
-def GetUnicodeString(string):
-  """Converts the string to Unicode if necessary."""
-  if not isinstance(string, py2to3.UNICODE_TYPE):
-    return str(string).decode('utf8', errors='ignore')
-  return string
+def GetUnicodeString(value):
+  """Attempts to convert the argument to a Unicode string.
+  
+  Args:
+    value (list|int|bytes): value to convert.
```
Comment (joachimmetz) [inline plaso/lib/objectfilter.py]: add `|str`

File: plaso/lib/utils.py (hunk 2 above)
Comment (joachimmetz) [reply plaso/lib/utils.py]: Ack, keep it as-is. I think `bytes_in` is what initially confused me.

**Comment by joachimmetz** [comment]
Some linter issues as well: https://travis-ci.org/log2timeline/plaso/jobs/390859521

**Comment by joachimmetz** [comment]
> "byte stream" is still strange to me - see https://en.wikipedia.org/wiki/Stream_(computing), which
> mentions in part that "a stream is a sequence of data elements made available over time". This matches
> my understanding, whereas the data in the "streams" mentioned in binary.py aren't made available over
> time, but all at once.

A different Wikipedia article says: https://en.wikipedia.org/wiki/Bitstream

```
A bytestream is a sequence of bytes. 
```

This says nothing about the availability of the bytes. Then again https://en.wikipedia.org/wiki/Reliable_byte_stream says:

```
A reliable byte stream is a common service paradigm
```

Based on that I would say yeah to more ambiguity ;)

If I look at:
https://docs.oracle.com/javase/tutorial/essential/io/bytestreams.html
https://msdn.microsoft.com/en-us/library/system.io.memorystream(v=vs.110).aspx

I would conclude that the term seems to be generally used for a sequence of bytes, backed by what ever form of medium or availability.

> I'd prefer to remove the usages of "stream" here, but you seem to feel strongly the other way.

I opt to do this in a different CL if you feel strongly about this. As I indicated, this code is likely to go away anyway with dtFabric. But please use something that is not `string` ;)

File: plaso/lib/objectfilter.py (hunk 6 above)
Comment (Onager) [reply plaso/lib/objectfilter.py]: Done

File: plaso/lib/binary.py [hunk 7]
```diff
@@ -211,31 +210,76 @@ def ArrayOfUTF16StreamCopyToStringTable(byte_stream, byte_stream_size=None):
...
+  # pylint: disable=no-member
+  if py2to3.PY_2:
+    iterators = itertools.izip_longest(*string_iterators)
+  else:
+    iterators = itertools.zip_longest(*string_iterators)
+  groups = list(iterators)
+  output_string = ''.join(
+      ['\\x{0:s}{1:s}'.format(group[0], group[1]) for group in groups])
+  return output_string
+
```
[the remark this answers is not in the document]
Comment (Onager) [reply plaso/lib/binary.py]: Done

File: plaso/lib/binary.py (hunk 5 above)
Comment (Onager) [reply plaso/lib/binary.py]: Split() is only faster for large strings, I think this is better as partition()

**Comment by joachimmetz** [comment]
Feel free to merge once tests pass

**Pull request merged**